plugins/multifundchannel.c

```diff
@@ -0,0 +1,1939 @@
+
+	If some interruption or problem occurs between steps 3
+	and 4, then the transaction is already in some node
+	mempool and will likely be broadcast, but `sendpsbt` has
+	failed.
+
+	And so we have to mark the channels as being "done"
+	*before* we do `sendpsbt`.
+	If not, if we error on `sendpsbt`, that would cause us to
+	`fundchannel_cancel` all the peers, but that is risky,
+	as, the funding transaction could still have been
+	broadcast and the channels funded.
+
+	That is, we treat `sendpsbt` failure as a possible
+	false negative.
+	*/
+	for (size_t i = 0; i < tal_count(mfc->destinations); ++i) {
+		struct multifundchannel_destination *dest;
+		dest = &mfc->destinations[i];
+		dest->state = MULTIFUNDCHANNEL_DONE;
+	}
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64": sendpsbt.", mfc->id);
+
+	req = jsonrpc_request_start(mfc->cmd->plugin, mfc->cmd,
+				    "sendpsbt",
+				    &after_sendpsbt,
+				    &mfc_forward_error,
+				    mfc);
+	json_add_psbt(req->js, "psbt", mfc->psbt);
+	return send_outreq(mfc->cmd->plugin, req);
+}
+
+static struct command_result *
+multifundchannel_finished(struct multifundchannel_command *mfc);
+
+static struct command_result *
+after_sendpsbt(struct command *cmd,
+	       const char *buf,
+	       const jsmntok_t *result,
+	       struct multifundchannel_command *mfc)
+{
+	const jsmntok_t *tx_tok;
+	const jsmntok_t *txid_tok;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64": sendpsbt done.", mfc->id);
+
+	tx_tok = json_get_member(buf, result, "tx");
+	if (!tx_tok)
+		plugin_err(cmd->plugin,
+			   "sendpsbt response has no 'tx': %.*s",
+			   json_tok_full_len(result),
+			   json_tok_full(buf, result));
+	mfc->final_tx = json_strdup(mfc, buf, tx_tok);
+
+	txid_tok = json_get_member(buf, result, "txid");
+	if (!txid_tok)
+		plugin_err(cmd->plugin,
+			   "sendpsbt response has no 'txid': %.*s",
+			   json_tok_full_len(result),
+			   json_tok_full(buf, result));
+	mfc->final_txid = json_strdup(mfc, buf, txid_tok);
+
+	/* PSBT is no longer something we need to clean up.  */
+	mfc->psbt = tal_free(mfc->psbt);
+
+	return multifundchannel_finished(mfc);
+}
+
+/*---------------------------------------------------------------------------*/
+/*~
+And finally we are done, after a thousand lines or so of code.
+*/
+
+static struct command_result *
+multifundchannel_finished(struct multifundchannel_command *mfc)
+{
+	unsigned int i;
+	struct json_stream *out;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %"PRIu64": done.", mfc->id);
+
+	out = jsonrpc_stream_success(mfc->cmd);
+	json_add_string(out, "tx", mfc->final_tx);
+	json_add_string(out, "txid", mfc->final_txid);
+	json_array_start(out, "channel_ids");
+	for (i = 0; i < tal_count(mfc->destinations); ++i) {
+		json_object_start(out, NULL);
+		json_add_node_id(out, "id", &mfc->destinations[i].id);
+		json_add_string(out, "channel_id", mfc->destinations[i].channel_id);
+		json_add_num(out, "outnum", mfc->destinations[i].outnum);
```

Comment:
```suggestion
		json_add_num(out, "funding_outnum", mfc->destinations[i].outnum);
```
This `outnum` is the `funding_tx`'s outnum